OPHD/UI/StringTable.cpp

```diff
@@ -0,0 +1,237 @@
+#include "StringTable.h"
+#include "../FontManager.h"
+#include "../Constants/UiConstants.h"
+#include <NAS2D/Utility.h>
+#include <stdexcept>
+
+const NAS2D::Color StringTable::Cell::ColorEmpty = NAS2D::Color::NoAlpha;
+
+StringTable::StringTable(std::size_t columns, std::size_t rows) : columnCount(columns), rowCount(rows)
+{
+	cells.resize(columns * rows);
+
+	defaultFont = NAS2D::Utility<FontManager>::get().font(constants::FONT_PRIMARY, constants::FONT_PRIMARY_NORMAL);
+	defaultTitleFont = NAS2D::Utility<FontManager>::get().font(constants::FONT_PRIMARY_BOLD, constants::FONT_PRIMARY_NORMAL);
+}
+
+void StringTable::draw(NAS2D::Renderer& renderer) const
+{
+	for (std::size_t i = 0; i < cells.size(); ++i)
+	{
+		const Cell& cell = cells.at(i);
+
+		NAS2D::Color textColor = cell.textColor != Cell::ColorEmpty ? cell.textColor : defaultTextColor;
+		const auto drawPosition = position + cell.textRelativePosition;
+
+		renderer.drawText(*getCellFont(i), cell.text, NAS2D::Point<float>(drawPosition.x, drawPosition.y), textColor);
+	}
+}
+
+void StringTable::setPosition(NAS2D::Vector<float> tablePosition)
```

Comment:
Would it be preferred to make all these short get/set functions as inline in the header file?

Reply:
Uncertain.

Sometimes it feels lazy to define methods in the header file. Under some conditions that can lead to duplicate code generation. On the other hand, having inline methods can aid inlining as a compiler optimization. I'm not sure if recent compilers need inline code to inline a method though. I believe such a transformation might fall under link time optimization, which is something older compilers didn't really have.

In this particular case, it might make sense to get the method by deriving from `Control`.

Reply:
Lets just leave the definitions in the cpp file for now. I suspect sense we are using C++17 compliant compilers, it is probably already optimizing well and no reason to try and push the compiler for inlining if there isn't a specific reason.'

I used to try to inline all the smaller calls, but I sort of like the cleaner header feel if you don't see inline functions. It makes it easier to read through the different function and variable names involved in the class. (I still prefer C# where you do not have to duplicate the prototype and definition for every class, but oh well.)

Reply:
See comment above about control interface.